src/coreclr/src/vm/ipcstreamfactory.cpp

```diff
@@ -0,0 +1,213 @@
+    {
+        // cache is empty, reconnect, e.g., there was a disconnect
+        IpcStream *pConnection = _pIpc->Connect(callback);
+        if (pConnection == nullptr)
+        {
+            if (callback != nullptr)
+                callback("Failed to connect to client connection", -1);
+            return false;
+        }
+        if (!DiagnosticsIpc::SendIpcAdvertise_V1(pConnection))
+        {
+            if (callback != nullptr)
+                callback("Failed to send advertise message", -1);
+            delete pConnection;
+            return false;
+        }
+
+        _pStream = pConnection;
+    }
+    *pIpcPollHandle = { nullptr, _pStream, 0, this };
+    return true;
+}
+
+IpcStream *IpcStreamFactory::ClientConnectionState::GetConnectedStream(ErrorCallback callback)
+{
+    IpcStream *pStream = _pStream;
+    _pStream = nullptr;
+    return pStream;
+}
+
+void IpcStreamFactory::ClientConnectionState::Reset(ErrorCallback callback)
+{
+    delete _pStream;
+    _pStream = nullptr;
+}
+
+bool IpcStreamFactory::ServerConnectionState::GetIpcPollHandle(IpcStream::DiagnosticsIpc::IpcPollHandle *pIpcPollHandle, ErrorCallback callback)
+{
+    *pIpcPollHandle = { _pIpc, nullptr, 0, this };
+    return true;
+}
+
+IpcStream *IpcStreamFactory::ServerConnectionState::GetConnectedStream(ErrorCallback callback)
+{
+    return _pIpc->Accept(callback);
+}
+
+// noop for server
+void IpcStreamFactory::ServerConnectionState::Reset(ErrorCallback)
+{
+    return;
+}
+
+bool IpcStreamFactory::CreateServer(const char *const pIpcName, ErrorCallback callback)
+{
+    IpcStream::DiagnosticsIpc *pIpc = IpcStream::DiagnosticsIpc::Create(pIpcName, IpcStream::DiagnosticsIpc::ConnectionMode::SERVER, callback);
+    if (pIpc != nullptr)
+    {
+        if (pIpc->Listen(callback))
+        {
+            s_rgpConnectionStates.Push(new ServerConnectionState(pIpc));
+            return true;
+        }
+        else
+        {
+            delete pIpc;
+            return false;
+        }
+    }
+    else
+    {
+        return false;
+    }
+}
+
+bool IpcStreamFactory::CreateClient(const char *const pIpcName, ErrorCallback callback)
+{
+    IpcStream::DiagnosticsIpc *pIpc = IpcStream::DiagnosticsIpc::Create(pIpcName, IpcStream::DiagnosticsIpc::ConnectionMode::CLIENT, callback);
+    if (pIpc != nullptr)
+    {
+        s_rgpConnectionStates.Push(new ClientConnectionState(pIpc));
+        return true;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+bool IpcStreamFactory::HasActiveConnections()
+{
+    return !s_isShutdown && s_rgpConnectionStates.Size() > 0;
+}
+
+void IpcStreamFactory::CloseConnections(ErrorCallback callback)
+{
+    s_isShutdown = true;
+    for (uint32_t i = 0; i < (uint32_t)s_rgpConnectionStates.Size(); i++)
+        s_rgpConnectionStates[i]->Close(callback);
+}
+
+void IpcStreamFactory::Shutdown(ErrorCallback callback)
+{
+    if (s_isShutdown)
+        return;
+    s_isShutdown = true;
+    s_ioEvent.Set();
```

Comment:
What is the required cleanup that the `s_ioEvent` triggers? In other words, if you delete the `s_ioEvent.Set();` line, what is not going to work?

Reply:
Everything should still work and shutdown shouldn't be blocked.  I've tested it locally and tests are passing when I have the call to [`IpcStreamFactory::Shutdown()`](https://github.com/dotnet/runtime/pull/35850/files#diff-677ac9d9fad5cae3b9a3d420d2f47d30R227) in `diagnosticserver.cpp` commented out.  The `s_ioEvent` is used strictly to unblock any ongoing calls to `IpcStreamFactory::Poll`.  My intent is to prevent regressing #12991 which was fixed by closing open handles to break the `Accept` calls that used to be where `Poll` is now.  I'm in the process of testing to see if that is necessary still or nor.